src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/DeferredProjectsUtility/DeferredProjectVSUtility.cs

```diff
@@ -0,0 +1,109 @@
+
+                // check if solution is DPL enabled or not. 
+                if (!IsSolutionDPLEnabled())
+                {
+                    return false;
+                }
+
+                // Get deferred projects count of current solution
+                var value = GetVSSolutionProperty((int)(__VSPROPID7.VSPROPID_DeferredProjectCount));
+                return (int)value != 0;
+            });
+        }
+
+        public static bool IsSolutionDPLEnabled()
+        {
+            return NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
+            {
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var vsSolution7 = _vsSolution.Value as IVsSolution7;
+
+                if (vsSolution7 != null && vsSolution7.IsSolutionLoadDeferred())
+                {
+                    return true;
+                }
+
+                return false;
+            });
+        }
+
+        public static async Task<IEnumerable<string>> GetDeferredProjectsFilePathAsync()
+        {
+            return await NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var projectPaths = new List<string>();
+                IEnumHierarchies enumHierarchies;
+                var guid = Guid.Empty;
+                var hr = _vsSolution.Value.GetProjectEnum((uint)__VSENUMPROJFLAGS3.EPF_DEFERRED, ref guid, out enumHierarchies);
+
+                ErrorHandler.ThrowOnFailure(hr);
+
+                // Loop all projects found
+                if (enumHierarchies != null)
+                {
+                    // Loop projects found
+                    var hierarchy = new IVsHierarchy[1];
+                    uint fetched = 0;
+                    while (enumHierarchies.Next(1, hierarchy, out fetched) == VSConstants.S_OK && fetched == 1)
+                    {
+                        string projectPath;
+                        hierarchy[0].GetCanonicalName(VSConstants.VSITEMID_ROOT, out projectPath);
+
+                        if (!string.IsNullOrEmpty(projectPath))
+                        {
+                            projectPaths.Add(projectPath);
+                        }
+                    }
+                }
+
+                return projectPaths;
+            });
+        }
+
+        private static object GetVSSolutionProperty(int propId)
+        {
+            Debug.Assert(ThreadHelper.CheckAccess());
+
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            object value;
+            var hr = _vsSolution.Value.GetProperty(propId, out value);
+
+            ErrorHandler.ThrowOnFailure(hr);
+
+            return value;
+        }
+
+        private static IVsSolution GetVsSolution()
+        {
+            var serviceProvider = ServiceLocator.GetInstance<IServiceProvider>();
```

Comment:
Use of `ServiceLocator` is not recommended as it may lead to a deadlock.
To guard against it add `ThreadHelper.ThrowIfNotOnUIThread();`